frame/balances/src/lib.rs

```diff
@@ -306,40 +277,91 @@ decl_error! {
+	Misc = 1,
+	/// Any reason at all.
+	All = 2,
 }
 
-impl<Balance: SimpleArithmetic + Copy, BlockNumber: SimpleArithmetic + Copy> VestingSchedule<Balance, BlockNumber> {
-	/// Amount locked at block `n`.
-	pub fn locked_at(&self, n: BlockNumber) -> Balance
-		where Balance: From<BlockNumber>
-	{
-		// Number of blocks that count toward vesting
-		// Saturating to 0 when n < starting_block
-		let vested_block_count = n.saturating_sub(self.starting_block);
-		// Return amount that is still locked in vesting
-		if let Some(x) = Balance::from(vested_block_count).checked_mul(&self.per_block) {
-			self.locked.max(x) - x
+impl From<WithdrawReasons> for Reasons {
+	fn from(r: WithdrawReasons) -> Reasons {
+		if r == WithdrawReasons::from(WithdrawReason::TransactionPayment) {
+			Reasons::Fee
+		} else if r.contains(WithdrawReason::TransactionPayment) {
+			Reasons::All
 		} else {
-			Zero::zero()
+			Reasons::Misc
 		}
 	}
 }
 
+impl BitOr for Reasons {
+	type Output = Reasons;
+	fn bitor(self, other: Reasons) -> Reasons {
+		if self == other { return self }
+		Reasons::All
+	}
+}
+
+/// A single lock on a balance. There can be many of these on an account and they "overlap", so the
+/// same balance is frozen by multiple locks.
 #[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
-pub struct BalanceLock<Balance, BlockNumber> {
+pub struct BalanceLock<Balance> {
+	/// An identifier for this lock. Only one lock may be in existence for each identifier.
 	pub id: LockIdentifier,
+	/// The amount which the free balance may not drop below when this lock is in effect.
 	pub amount: Balance,
-	pub until: BlockNumber,
-	pub reasons: WithdrawReasons,
+	/// If true, then the lock remains in effect even for payment of transaction fees.
+	pub reasons: Reasons,
+}
+
+/// All balance information for an account.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug)]
+pub struct AccountData<Balance> {
+	/// Non-reserved part of the balance. There may still be restrictions on this, but it is the
+	/// total pool what may in principle be transferred, reserved and used for tipping.
+	///
+	/// This is the only balance that matters in terms of most operations on tokens. It
+	/// alone is used to determine the balance when in the contract execution environment.
+	pub free: Balance,
+	/// Balance which is reserved and may not be used at all.
+	///
+	/// This can still get slashed, but gets slashed last of all.
+	///
+	/// This balance is a 'reserve' balance that other subsystems use in order to set aside tokens
+	/// that are still 'owned' by the account holder, but which are suspendable.
+	pub reserved: Balance,
+	/// The amount that `free` may not drop below when withdrawing for *anything except transaction
+	/// fee payment*.
+	pub misc_frozen: Balance,
+	/// The amount that `free` may not drop below when withdrawing specifically for transaction
+	/// fee payment.
+	pub fee_frozen: Balance,
```

Comment:
Just making a comment here while I dive into the code. It is not immediately clear to me the story around `misc_frozen` and `fee_frozen`.

A user will have a list of locks, each with a different id, amount, and a set of reasons.

The largest lock with reason ONLY for `TransactionPayment` will determine the `fee_frozen`.

The largest lock without `TransactionPayment` as the reason will determine the `misc_frozen`.

If a lock has some combination of reasons that include `TransactionPayment` and something else, we use the greatest of the two frozen values.

So I guess the question in my head is what is the most obvious scenario for having `misc_frozen` and `fee_frozen` be different values?

I assume `fee_frozen` would on average be less than `misc_frozen`, which would allow a user to use some balance specifically for fees, when they would otherwise be locked for other reasons?

What kinds of situations would this occur?

Reply:
Yes. Locks will typically be for: everything except fees (because we still want to let users pay for transactions with the locked funds - to unlock them for example); and everything (in the case where we don't even want them to pay fees with the locked funds). So basically `Reasons::Misc` and `Reasons::All`. It's very unlikely that `Reasons::Fees` would ever be used, but the code was simpler this way and it's strictly more flexible so I left it.